// ngx-scrollbar/esm2022/lib/async-detection.mjs
import { Directive, Input, inject, NgZone } from '@angular/core';
import { NgScrollbarExt } from './ng-scrollbar-ext';
import { mutationObserver } from './viewport';
import * as i0 from "@angular/core";
export class AsyncDetection {
    constructor() {
        this.scrollbar = inject(NgScrollbarExt);
        this.zone = inject(NgZone);
        this.scrollbar.skipInit = true;
    }
    ngOnInit() {
        this.zone.runOutsideAngular(() => {
            this.subscription = mutationObserver(this.scrollbar.nativeElement, 100).subscribe(() => {
                if (!this.scrollbar.viewport.initialized()) {
                    // Search for external viewport
                    const viewportElement = this.scrollbar.nativeElement.querySelector(this.scrollbar.externalViewport);
                    // Search for external content wrapper
                    const contentWrapperElement = this.scrollbar.nativeElement.querySelector(this.scrollbar.externalContentWrapper);
                    if (viewportElement && contentWrapperElement) {
                        // If an external spacer selector is provided, search for it
                        let spacerElement;
                        if (this.scrollbar.externalSpacer) {
                            spacerElement = this.scrollbar.nativeElement.querySelector(this.scrollbar.externalSpacer);
                        }
                        // Initialize viewport
                        this.scrollbar.viewport.init(viewportElement, contentWrapperElement, spacerElement);
                        // Attach scrollbars
                        this.scrollbar.attachScrollbars();
                        if (!this.asyncDetection) {
                            this.subscription.unsubscribe();
                        }
                    }
                }
                else {
                    const viewportElement = this.scrollbar.nativeElement.querySelector(this.scrollbar.externalViewport);
                    const contentWrapperElement = this.scrollbar.nativeElement.querySelector(this.scrollbar.externalContentWrapper);
                    if (!viewportElement || !contentWrapperElement) {
                        this.scrollbar.viewport.nativeElement = null;
                        this.scrollbar.viewport.contentWrapperElement = null;
                        this.scrollbar.viewport.initialized.set(false);
                    }
                }
            });
        });
    }
    ngOnDestroy() {
        this.subscription?.unsubscribe();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: AsyncDetection, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.2", type: AsyncDetection, isStandalone: true, selector: "ng-scrollbar[externalViewport][asyncDetection]", inputs: { asyncDetection: "asyncDetection" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: AsyncDetection, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-scrollbar[externalViewport][asyncDetection]'
                }]
        }], ctorParameters: () => [], propDecorators: { asyncDetection: [{
                type: Input
            }] } });
