// ngx-scrollbar/esm2022/reached-event/ng-scroll-dropped.mjs
import { Directive, Input, Output, numberAttribute, booleanAttribute, input, EventEmitter } from '@angular/core';
import { ReachedDroppedBase } from './reached-dropped-base';
import * as i0 from "@angular/core";
export class NgScrollDropped extends ReachedDroppedBase {
    constructor() {
        super(...arguments);
        this.disabled = input(false, {
            alias: 'disableDropped',
            transform: booleanAttribute
        });
        this.top = new EventEmitter();
        this.bottom = new EventEmitter();
        this.start = new EventEmitter();
        this.end = new EventEmitter();
        this.triggerElementsWrapperClass = 'ng-scroll-dropped-wrapper';
        this.triggerElementClass = 'scroll-dropped-trigger-element';
    }
    /** Dropped offset value in px */
    set offsetSetter(value) {
        this.setCssVariable('--dropped-offset', value);
    }
    set offsetTopSetter(value) {
        this.setCssVariable('--dropped-offset-top', value);
    }
    set offsetBottomSetter(value) {
        this.setCssVariable('--dropped-offset-bottom', value);
    }
    set offsetStartSetter(value) {
        this.setCssVariable('--dropped-offset-start', value);
    }
    set offsetEndSetter(value) {
        this.setCssVariable('--dropped-offset-end', value);
    }
    isTriggered(entry) {
        return !entry.isIntersecting;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NgScrollDropped, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.1.2", type: NgScrollDropped, isStandalone: true, selector: "ng-scrollbar[droppedTop], ng-scrollbar[droppedBottom], ng-scrollbar[droppedStart], ng-scrollbar[droppedEnd]", inputs: { offsetSetter: { classPropertyName: "offsetSetter", publicName: "droppedOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetTopSetter: { classPropertyName: "offsetTopSetter", publicName: "droppedTopOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetBottomSetter: { classPropertyName: "offsetBottomSetter", publicName: "droppedBottomOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetStartSetter: { classPropertyName: "offsetStartSetter", publicName: "droppedStartOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetEndSetter: { classPropertyName: "offsetEndSetter", publicName: "droppedEndOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, disabled: { classPropertyName: "disabled", publicName: "disableDropped", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { top: "droppedTop", bottom: "droppedBottom", start: "droppedStart", end: "droppedEnd" }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NgScrollDropped, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-scrollbar[droppedTop], ng-scrollbar[droppedBottom], ng-scrollbar[droppedStart], ng-scrollbar[droppedEnd]',
                    standalone: true,
                }]
        }], propDecorators: { offsetSetter: [{
                type: Input,
                args: [{ alias: 'droppedOffset', transform: numberAttribute }]
            }], offsetTopSetter: [{
                type: Input,
                args: [{ alias: 'droppedTopOffset', transform: numberAttribute }]
            }], offsetBottomSetter: [{
                type: Input,
                args: [{ alias: 'droppedBottomOffset', transform: numberAttribute }]
            }], offsetStartSetter: [{
                type: Input,
                args: [{ alias: 'droppedStartOffset', transform: numberAttribute }]
            }], offsetEndSetter: [{
                type: Input,
                args: [{ alias: 'droppedEndOffset', transform: numberAttribute }]
            }], top: [{
                type: Output,
                args: ['droppedTop']
            }], bottom: [{
                type: Output,
                args: ['droppedBottom']
            }], start: [{
                type: Output,
                args: ['droppedStart']
            }], end: [{
                type: Output,
                args: ['droppedEnd']
            }] } });
