// ngx-scrollbar/esm2022/reached-event/ng-scroll-reached.mjs
import { Directive, Input, Output, numberAttribute, booleanAttribute, input, EventEmitter } from '@angular/core';
import { ReachedDroppedBase } from './reached-dropped-base';
import * as i0 from "@angular/core";
export class NgScrollReached extends ReachedDroppedBase {
    constructor() {
        super(...arguments);
        this.disabled = input(false, {
            alias: 'disableReached',
            transform: booleanAttribute
        });
        this.top = new EventEmitter();
        this.bottom = new EventEmitter();
        this.start = new EventEmitter();
        this.end = new EventEmitter();
        this.triggerElementsWrapperClass = 'ng-scroll-reached-wrapper';
        this.triggerElementClass = 'scroll-reached-trigger-element';
    }
    /** Reached offset value in px */
    set offsetSetter(value) {
        this.setCssVariable('--reached-offset', value);
    }
    set offsetTopSetter(value) {
        this.setCssVariable('--reached-offset-top', value);
    }
    set offsetBottomSetter(value) {
        this.setCssVariable('--reached-offset-bottom', value);
    }
    set offsetStartSetter(value) {
        this.setCssVariable('--reached-offset-start', value);
    }
    set offsetEndSetter(value) {
        this.setCssVariable('--reached-offset-end', value);
    }
    isTriggered(entry) {
        return entry.isIntersecting;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NgScrollReached, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.1.2", type: NgScrollReached, isStandalone: true, selector: "ng-scrollbar[reachedTop], ng-scrollbar[reachedBottom], ng-scrollbar[reachedStart], ng-scrollbar[reachedEnd]", inputs: { offsetSetter: { classPropertyName: "offsetSetter", publicName: "reachedOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetTopSetter: { classPropertyName: "offsetTopSetter", publicName: "reachedTopOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetBottomSetter: { classPropertyName: "offsetBottomSetter", publicName: "reachedBottomOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetStartSetter: { classPropertyName: "offsetStartSetter", publicName: "reachedStartOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, offsetEndSetter: { classPropertyName: "offsetEndSetter", publicName: "reachedEndOffset", isSignal: false, isRequired: false, transformFunction: numberAttribute }, disabled: { classPropertyName: "disabled", publicName: "disableReached", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { top: "reachedTop", bottom: "reachedBottom", start: "reachedStart", end: "reachedEnd" }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.2", ngImport: i0, type: NgScrollReached, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-scrollbar[reachedTop], ng-scrollbar[reachedBottom], ng-scrollbar[reachedStart], ng-scrollbar[reachedEnd]',
                    standalone: true,
                }]
        }], propDecorators: { offsetSetter: [{
                type: Input,
                args: [{ alias: 'reachedOffset', transform: numberAttribute }]
            }], offsetTopSetter: [{
                type: Input,
                args: [{ alias: 'reachedTopOffset', transform: numberAttribute }]
            }], offsetBottomSetter: [{
                type: Input,
                args: [{ alias: 'reachedBottomOffset', transform: numberAttribute }]
            }], offsetStartSetter: [{
                type: Input,
                args: [{ alias: 'reachedStartOffset', transform: numberAttribute }]
            }], offsetEndSetter: [{
                type: Input,
                args: [{ alias: 'reachedEndOffset', transform: numberAttribute }]
            }], top: [{
                type: Output,
                args: ['reachedTop']
            }], bottom: [{
                type: Output,
                args: ['reachedBottom']
            }], start: [{
                type: Output,
                args: ['reachedStart']
            }], end: [{
                type: Output,
                args: ['reachedEnd']
            }] } });
